frontend/exporter/src/types/copied.rs

```diff
@@ -67,6 +77,380 @@ impl<'tcx, S: UnderOwnerState<'tcx>> SInto<S, GlobalIdent> for rustc_hir::def_id
+    },
+    Const {
+        #[value(s.base().tcx.generics_of(s.owner_id()).sinto(s))]
+        generics: TyGenerics,
+        #[value(get_generic_predicates(s, s.owner_id()))]
+        predicates: GenericPredicates,
+        #[value(s.base().tcx.type_of(s.owner_id()).instantiate_identity().sinto(s))]
+        ty: Ty,
+    },
+    /// Associated constant: `trait MyTrait { const ASSOC: usize; }`
+    AssocConst {
+        #[value(s.base().tcx.parent(s.owner_id()).sinto(s))]
+        parent: DefId,
+        #[value(s.base().tcx.associated_item(s.owner_id()).sinto(s))]
+        associated_item: AssocItem,
+        #[value(s.base().tcx.generics_of(s.owner_id()).sinto(s))]
+        generics: TyGenerics,
+        #[value(get_generic_predicates(s, s.owner_id()))]
+        predicates: GenericPredicates,
+        #[value(s.base().tcx.type_of(s.owner_id()).instantiate_identity().sinto(s))]
+        ty: Ty,
+    },
+    Static {
+        /// Whether it's a `unsafe static`, `safe static` (inside extern only) or just a `static`.
+        safety: Safety,
+        /// Whether it's a `static mut` or just a `static`.
+        mutability: Mutability,
+        /// Whether it's an anonymous static generated for nested allocations.
+        nested: bool,
+        #[value(s.base().tcx.generics_of(s.owner_id()).sinto(s))]
+        generics: TyGenerics,
+        #[value(get_generic_predicates(s, s.owner_id()))]
+        predicates: GenericPredicates,
+        #[value(s.base().tcx.type_of(s.owner_id()).instantiate_identity().sinto(s))]
+        ty: Ty,
+    },
+    /// Constant generic parameter: `struct Foo<const N: usize> { ... }`
+    ConstParam,
+    /// Refers to the struct or enum variant's constructor.
+    Ctor(CtorOf, CtorKind),
+
+    // Macro namespace
+    Macro(MacroKind),
+
+    // Not namespaced (or they are, but we don't treat them so)
+    ExternCrate,
+    Use,
+    /// An `extern` block.
+    ForeignMod,
+    /// Anonymous constant, e.g. the `1 + 2` in `[u8; 1 + 2]`
+    AnonConst,
+    /// An inline constant, e.g. `const { 1 + 2 }`
+    InlineConst,
+    /// Opaque type, aka `impl Trait`.
+    OpaqueTy,
+    Impl {
+        #[value(s.base().tcx.generics_of(s.owner_id()).sinto(s))]
+        generics: TyGenerics,
+        #[value(get_generic_predicates(s, s.owner_id()))]
+        predicates: GenericPredicates,
+        #[value(s.base().tcx.impl_subject(s.owner_id()).instantiate_identity().sinto(s))]
+        impl_subject: ImplSubject,
+        /// Associated items, in definition order.
+        #[value(s.base().tcx.associated_items(s.owner_id()).in_definition_order().collect::<Vec<_>>().sinto(s))]
+        items: Vec<AssocItem>,
+    },
+    /// A field in a struct, enum or union. e.g.
+    /// - `bar` in `struct Foo { bar: u8 }`
+    /// - `Foo::Bar::0` in `enum Foo { Bar(u8) }`
+    Field,
+    /// Lifetime parameter: the `'a` in `struct Foo<'a> { ... }`
+    LifetimeParam,
+    /// A use of `global_asm!`.
+    GlobalAsm,
+}
+
+impl FullDef {
```

Comment:
Seeing how lengthy things are, I wonder if we should not move `FullDefKind` and impls on it to its own file in `types/new`.

Reply:
Good idea